# Guided Project: Exploring Hacker News Posts

[Hacker News](https://news.ycombinator.com/) is a site where user-submitted stories are voted and commented upon, similar to reddit. The main topics for discussions are IT, Computer Security and startups.

We will explore the [Kaggles data set](https://www.kaggle.com/hacker-news/hacker-news-posts) with approximately 20,000 rows. Data set contains information about posts: title, url, number of points and comments. During our analysis the two types of posts will be compared (Ask HN and Show HN):

- Do Ask HN or Show HN receive more comments on average?
- Do posts created at a certain time receive more comments on average?

## Introduction

Let's import the data set and separate the headers.

In [1]:
from csv import reader

opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
headers = hn[0]
hn = hn[1:]

In [2]:
headers

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

In [3]:
hn[:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

## Extracting Ask HN and Show HN Posts

Now we are ready to process the data and create three lists to store "Ask", "Show" and other posts separately.

In [4]:
ask_posts, show_posts, other_posts = [], [], []

for row in hn:
    lower_title = row[1].lower()
    
    if lower_title.startswith('ask hn'):
        ask_posts.append(row)
    elif lower_title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [5]:
len_ask = len(ask_posts)
len_show = len(show_posts)
len_other = len(other_posts)
len_sum = len_ask + len_show + len_other

per_ask = round((len_ask / len_sum) * 100, 2) 
per_show = round((len_show / len_sum) * 100, 2)
per_other = round((len_other / len_sum) * 100, 2)

print('Ask posts: {} ({}%)'.format(len_ask, per_ask))
print('Show posts: {} ({}%)'.format(len_show, per_show))
print('Other posts: {} ({})%'.format(len_other, per_other))

Ask posts: 1744 (8.68%)
Show posts: 1162 (5.78%)
Other posts: 17194 (85.54)%


As we can see, "ask" posts are more frequent (8.68%) than "show" posts (5.78%). However, this is not the main site path.

## Calculating the Average Number of Comments for Ask HN and Show HN Posts

Let's determine if ask posts or show posts receive more comments on average.

In [6]:
ask_comments = [int(row[4]) for row in ask_posts]
show_comments = [int(row[4]) for row in show_posts]
avg_ask_comments = sum(ask_comments) / len(ask_comments)
avg_show_comments = sum(show_comments) / len(show_comments)    

In [7]:
print('Average number of comments for Ask HN:', avg_ask_comments)
print('Average number of comments for Show HN:', avg_show_comments)

Average number of comments for Ask HN: 14.038417431192661
Average number of comments for Show HN: 10.31669535283993


On average, ask posts receive approximately 14 comments, whereas show posts receive only 10.

## Finding the Amount of Ask Posts and Comments by Hour Created

Next, we will focus just on ask posts since they are more likely to receive comments. We will determine if ask posts created at certain time are more likely to attract comments.

In [12]:
time_comments = [(row[4], row[6]) for row in ask_posts]
time_comments[:5]

[('6', '8/16/2016 9:55'),
 ('29', '11/22/2015 13:43'),
 ('1', '5/2/2016 10:14'),
 ('3', '8/2/2016 14:20'),
 ('17', '10/15/2015 16:38')]

In [36]:
import datetime as dt

counts_by_hour = {}
comments_by_hour = {}

for n_comments, created_date in time_comments:
    n_comments = int(n_comments)
    created_dt = dt.datetime.strptime(created_date, '%m/%d/%Y %H:%M')
    created_hour = created_dt.hour
    
    if created_hour not in counts_by_hour:
        counts_by_hour[created_hour] = 1
    else:
        counts_by_hour[created_hour] += 1
    
    if created_hour not in comments_by_hour:
        comments_by_hour[created_hour] = n_comments
    else:
        comments_by_hour[created_hour] += n_comments

In [37]:
counts_by_hour

{0: 55,
 1: 60,
 2: 58,
 3: 54,
 4: 47,
 5: 46,
 6: 44,
 7: 34,
 8: 48,
 9: 45,
 10: 59,
 11: 58,
 12: 73,
 13: 85,
 14: 107,
 15: 116,
 16: 108,
 17: 100,
 18: 109,
 19: 110,
 20: 80,
 21: 109,
 22: 71,
 23: 68}

In [38]:
comments_by_hour

{0: 447,
 1: 683,
 2: 1381,
 3: 421,
 4: 337,
 5: 464,
 6: 397,
 7: 267,
 8: 492,
 9: 251,
 10: 793,
 11: 641,
 12: 687,
 13: 1253,
 14: 1416,
 15: 4477,
 16: 1814,
 17: 1146,
 18: 1439,
 19: 1188,
 20: 1722,
 21: 1745,
 22: 479,
 23: 543}

## Calculating the Average Number of Comments for Ask HN Posts by Hour

Now, we will use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.

In [39]:
avg_by_hour = {}

for hour in comments_by_hour:
    avg_by_hour[hour] = comments_by_hour[hour] / counts_by_hour[hour]
    
avg_by_hour

{0: 8.127272727272727,
 1: 11.383333333333333,
 2: 23.810344827586206,
 3: 7.796296296296297,
 4: 7.170212765957447,
 5: 10.08695652173913,
 6: 9.022727272727273,
 7: 7.852941176470588,
 8: 10.25,
 9: 5.5777777777777775,
 10: 13.440677966101696,
 11: 11.051724137931034,
 12: 9.41095890410959,
 13: 14.741176470588234,
 14: 13.233644859813085,
 15: 38.5948275862069,
 16: 16.796296296296298,
 17: 11.46,
 18: 13.20183486238532,
 19: 10.8,
 20: 21.525,
 21: 16.009174311926607,
 22: 6.746478873239437,
 23: 7.985294117647059}

## Sorting and Printing Values from a List of Lists

In [43]:
swap_avg_by_hour = []

for hour, avg in avg_by_hour.items():
    swap_avg_by_hour.append((avg, hour))
    
swap_avg_by_hour

[(8.127272727272727, 0),
 (11.383333333333333, 1),
 (23.810344827586206, 2),
 (7.796296296296297, 3),
 (7.170212765957447, 4),
 (10.08695652173913, 5),
 (9.022727272727273, 6),
 (7.852941176470588, 7),
 (10.25, 8),
 (5.5777777777777775, 9),
 (13.440677966101696, 10),
 (11.051724137931034, 11),
 (9.41095890410959, 12),
 (14.741176470588234, 13),
 (13.233644859813085, 14),
 (38.5948275862069, 15),
 (16.796296296296298, 16),
 (11.46, 17),
 (13.20183486238532, 18),
 (10.8, 19),
 (21.525, 20),
 (16.009174311926607, 21),
 (6.746478873239437, 22),
 (7.985294117647059, 23)]

In [47]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

[(38.5948275862069, 15),
 (23.810344827586206, 2),
 (21.525, 20),
 (16.796296296296298, 16),
 (16.009174311926607, 21)]

### Top 5 Hour for Ask Posts Comments

In [56]:
for c, h in sorted_swap[:5]:
    h = dt.time(hour=h).strftime('%H:%M')
    print('{}: {:.2f} average comments per post'.format(h, c))

15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The most popular hour for Ask posts is 15:00, with an average of 38.59 comments per post. This is about a 60% increase in the number of comments between the highest and second highest average number of comments.

## Conclusion

In this project we analyzed Hacker News posts to determine which type of post and time receive the most comments on average. Based on our analysis, the maximum amount of comments could be achieved using "Ask HN" posts with created time between 15:00 and 16:00 ET.